src/Persisters/Collection/OneToManyPersister.php

```diff
@@ -183,6 +183,12 @@ private function deleteEntityCollection(PersistentCollection $collection): int
             $types[]      = PersisterHelper::getTypeOfColumn($joinColumn['referencedColumnName'], $sourceClass, $this->em);
         }
 
+        if ($targetClass->discriminatorValue && $targetClass->discriminatorColumn) {
```

Comment:
I am not sure if this will work for joined table inheritance as well, can you use joined inheritance instead of single table in the test please? Single table is a special case of joined table

Reply:
This change doesn't affect joined inheritance, because its collections are removed using `deleteJoinedEntityCollection()` instead of `deleteEntityCollection()`

For joined inheritance this bug doesn't appear, because separate collections are filtered using JOIN to exact tables.

```
...
        $sql = $query->getSQL();
        //   SELECT c0_.id AS id_0 FROM Dog d1_
        //        INNER JOIN collection_with_inheritance_pet c0_ ON d1_.id = c0_.id WHERE d1_.petStore_id = ?

        assert(is_string($sql));
        $statement = 'INSERT INTO ' . $tempTable . ' (' . $idColumnList . ') ' . $sql;
        //   INSERT INTO collection_with_inheritance_pet_id_tmp (id)
        //        SELECT c0_.id AS id_0 FROM Dog d1_
        //            INNER JOIN collection_with_inheritance_pet c0_ ON d1_.id = c0_.id
        //        WHERE d1_.petStore_id = ?

        $parameters = array_values($sourceClass->getIdentifierValues($collection->getOwner()));
        $numDeleted = $this->conn->executeStatement($statement, $parameters);

        // 3) Delete records on each table in the hierarchy
        $classNames = array_merge($targetClass->parentClasses, [$targetClass->name], $targetClass->subClasses);

        foreach (array_reverse($classNames) as $className) {
            $tableName = $this->quoteStrategy->getTableName($this->em->getClassMetadata($className), $this->platform);
            $statement = 'DELETE FROM ' . $tableName . ' WHERE (' . $idColumnList . ')'
                . ' IN (SELECT ' . $idColumnList . ' FROM ' . $tempTable . ')';
            //   DELETE FROM collection_with_inheritance_pet
            //   WHERE (id) IN (SELECT id FROM collection_with_inheritance_pet_id_tmp)

            $this->conn->executeStatement($statement);
        }
...
```